remove.cpp

```diff
@@ -0,0 +1,211 @@
+/*
+ * Fledge "asset" filter plugin  remove rule
+ *
+ * Copyright (c) 2025 Dianomic Systems
+ *
+ * Released under the Apache 2.0 Licence
+ *
+ * Author: Mark Riddoch
+ */
+#include <logger.h>
+#include <rules.h>
+#include <set>
+
+using namespace std;
+using namespace rapidjson;
+
+static const    set<string> validDpTypes{"FLOAT", "INTEGER", "STRING", "FLOAT_ARRAY", "DP_DICT", "DP_LIST", "IMAGE", "DATABUFFER", "2D_FLOAT_ARRAY", "NUMBER", "NON-NUMERIC", "NESTED", "ARRAY", "2D_ARRAY", "USER_ARRAY"};
+
+/**
+ * Constructor for the remove rule
+ *
+ * @param asset	The asset name
+ * @param json	JSON object
+ */
+RemoveRule::RemoveRule(const string& asset, const rapidjson::Value& json) : Rule(asset), m_regex(NULL)
+{
+	if (json.HasMember("datapoint") && json["datapoint"].IsString())
+	{
+		string datapoint = json["datapoint"].GetString();
+		if (isRegexString(datapoint))
+			m_regex = new regex(datapoint);
+		else
+			m_datapoint = datapoint;
+	}
+	else if (json.HasMember("type") && json["type"].IsString())
+	{
+		m_type = json["type"].GetString();
+		transform(m_type.begin(), m_type.end(), m_type.begin(), ::toupper);
+		if (m_type == "FLOATING")
+			m_type = "FLOAT";
+		if (m_type == "BUFFER")
+			m_type = "DATABUFFER";
+		if (m_type == "NESTED")
+			m_type = "DP_DICT";
+		if (m_type == "2D_ARRAY")
+			m_type = "2D_FLOAT_ARRAY";
+		if (m_type == "ARRAY")
+			m_type = "FLOAT_ARRAY";
+
+		if (!validateType(m_type))
+		{
+			m_logger->warn("Invalid Datapoint type %s given in rule for asset '%s'. The rule will have no impact.", m_type.c_str(), m_asset.c_str());
+		}
+	}
+	else if (json.HasMember("datapoints") && json["datapoints"].IsArray())
+	{
+		const Value& dps = json["datapoints"];
+		for (auto& dp : dps.GetArray())
+		{
+			if (dp.IsString())
+				m_datapoints.push_back(dp.GetString());
+			else
+				m_logger->error("The datapoints in the array of names for the asset '%s' must all be strings.", m_asset.c_str());
+		}
+	}
+	else
+	{
+		m_logger->error("Badly defined remove rule for asset '%s', a 'datapoint', 'type' or 'datapoints' property must be given The 'datapoint' and 'type' properties must be strings and 'datapopints' is expected to be an array of strings.", m_asset.c_str());
```

Comment:
Suggest breaking this into multiple sentences because there are many possibilities in this error. How about "Badly defined Remove rule for asset '%s'. A 'datapoint', 'type' or 'datapoints' property must be given. The 'datapoint' and 'type' properties must be strings and 'datapoints' is expected to be an array of strings."